File: Darts/src/burlton/dartzee/code/db/AchievementEntity.java
package burlton.dartzee.code.db;

import burlton.core.code.obj.HandyArrayList;
import burlton.dartzee.code.achievements.AbstractAchievement;
import burlton.dartzee.code.achievements.AchievementUtilKt;
import burlton.dartzee.code.screen.ScreenCache;
import burlton.dartzee.code.screen.game.DartsGameScreen;
import burlton.desktopcore.code.util.DateUtil;

import java.awt.*;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.util.ArrayList;

/**
 * Entity to record a particular that a particular achievement has been earned by a player.
 * 
 * Points at a GameId if applicable, and uses DtCreation as the "date unlocked".
 */
public final class AchievementEntity extends AbstractEntity<AchievementEntity>
{
	private long playerId = -1;
	private int achievementRef = -1;
	private long gameIdEarned = -1;
	private int achievementCounter = -1;
	
	@Override
	public String getTableName()
	{
		return "Achievement";
	}

	@Override
	public String getCreateTableSqlSpecific()
	{
		return "PlayerId INT NOT NULL, "
		  + "AchievementRef INT NOT NULL, "
		  + "GameIdEarned INT NOT NULL, "
		  + "AchievementCounter INT NOT NULL";
	}

	@Override
	public void populateFromResultSet(AchievementEntity entity, ResultSet rs)
			throws SQLException
	{
		entity.setPlayerId(rs.getInt("PlayerId"));
		entity.setAchievementRef(rs.getInt("AchievementRef"));
		entity.setGameIdEarned(rs.getInt("GameIdEarned"));
		entity.setAchievementCounter(rs.getInt("AchievementCounter"));
	}

	@Override
	public String writeValuesToStatement(PreparedStatement statement, int i, String statementStr) throws SQLException
	{
		statementStr = writeLong(statement, i++, playerId, statementStr);
		statementStr = writeInt(statement, i++, achievementRef, statementStr);
		statementStr = writeLong(statement, i++, gameIdEarned, statementStr);
		statementStr = writeInt(statement, i++, achievementCounter, statementStr);
		
		return statementStr;
	}
	
	@Override
	public void addListsOfColumnsForIndexes(ArrayList<ArrayList<String>> indexes)
	{
		ArrayList<String> ix = HandyArrayList.factoryAdd("PlayerId", "AchievementRef");
		
		indexes.add(ix);
	}
	
	public static AchievementEntity retrieveAchievement(int achievementRef, long playerId)
	{
		return new AchievementEntity().retrieveEntity("PlayerId = " + playerId + " AND AchievementRef = " + achievementRef);
	}

	/**
	 *  Methods for gameplay logic to update achievements
	 */
	public static void updateAchievement(int achievementRef, long playerId, long gameId, int counter)
	{
		AchievementEntity existingAchievement = retrieveAchievement(achievementRef, playerId);
	
		if (existingAchievement == null)
		{
			AchievementEntity.factoryAndSave(achievementRef, playerId, gameId, counter);

			triggerAchievementUnlock(-1, counter, achievementRef, playerId, gameId);
		}
		else
		{
			int existingCounter = existingAchievement.getAchievementCounter();

			boolean decreasing = AchievementUtilKt.getAchievementForRef(achievementRef).isDecreasing();

			//Update the achievement if appropriate
			if ((counter > existingCounter && !decreasing)
			 || (counter < existingCounter && decreasing))
			{
				existingAchievement.setAchievementCounter(counter);
				existingAchievement.setGameIdEarned(gameId);
				existingAchievement.saveToDatabase();

				triggerAchievementUnlock(existingCounter, counter, achievementRef, playerId, gameId);
			}
		}
	}
	public static void incrementAchievement(int achievementRef, long playerId, long gameId, int amountBy)
	{
		AchievementEntity existingAchievement = retrieveAchievement(achievementRef, playerId);

		if (existingAchievement == null)
		{
			AchievementEntity.factoryAndSave(achievementRef, playerId, -1, amountBy);

			triggerAchievementUnlock(-1, amountBy, achievementRef, playerId, gameId);
		}
		else
		{
			int existingCount = existingAchievement.getAchievementCounter();
			existingAchievement.setAchievementCounter(existingCount + amountBy);
			existingAchievement.saveToDatabase();

			triggerAchievementUnlock(existingCount, existingCount + amountBy, achievementRef, playerId, gameId);
		}
	}
	private static void triggerAchievementUnlock(int oldValue, int newValue, int achievementRef, long playerId, long gameId)
	{
		AbstractAchievement achievementTemplate = AchievementUtilKt.getAchievementForRef(achievementRef);
		triggerAchievementUnlock(oldValue, newValue, achievementTemplate, playerId, gameId);
	}
	public static void triggerAchievementUnlock(int oldValue, int newValue, AbstractAchievement achievementTemplate, long playerId, long gameId)
	{
		//Work out if the threshold has changed
		achievementTemplate.setAttainedValue(oldValue);

		Color oldColor = achievementTemplate.getColor(false);

		achievementTemplate.setAttainedValue(newValue);
		Color newColor = achievementTemplate.getColor(false);

		if (oldColor != newColor)
		{
			//Hooray we've done a thing!
			DartsGameScreen scrn = ScreenCache.getDartsGameScreen(gameId);
			scrn.achievementUnlocked(gameId, playerId, achievementTemplate);
		}
	}
	
	public static AchievementEntity factoryAndSave(int achievementRef, long playerId, long gameId, int counter)
	{
		return factoryAndSave(achievementRef, playerId, gameId, counter, DateUtil.getSqlDateNow());
	}
	public static AchievementEntity factoryAndSave(int achievementRef, long playerId, long gameId, int counter,
	  Timestamp dtLastUpdate)
	{
		AchievementEntity ae = new AchievementEntity();
		ae.assignRowId();
		ae.achievementRef = achievementRef;
		ae.playerId = playerId;
		ae.gameIdEarned = gameId;
		ae.achievementCounter = counter;
		ae.saveToDatabase(dtLastUpdate);
		
		return ae;
	}

	/**
	 * Gets / Sets
	 */
	public long getPlayerId()
	{
		return playerId;
	}
	public void setPlayerId(long playerId)
	{
		this.playerId = playerId;
	}
	public int getAchievementRef()
	{
		return achievementRef;
	}
	public void setAchievementRef(int achievementId)
	{
		this.achievementRef = achievementId;
	}
	public long getGameIdEarned()
	{
		return gameIdEarned;
	}
	public void setGameIdEarned(long gameIdEarned)
	{
		this.gameIdEarned = gameIdEarned;
	}
	public int getAchievementCounter()
	{
		return achievementCounter;
	}
	public void setAchievementCounter(int achievementCounter)
	{
		this.achievementCounter = achievementCounter;
	}
}

File: Darts/src/burlton/dartzee/code/db/AchievementEntity.kt
package burlton.dartzee.code.db

import burlton.core.code.obj.HandyArrayList
import burlton.dartzee.code.achievements.AbstractAchievement
import burlton.dartzee.code.achievements.getAchievementForRef
import burlton.dartzee.code.screen.ScreenCache
import burlton.dartzee.code.utils.DatabaseUtil
import burlton.desktopcore.code.util.getSqlDateNow
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException
import java.sql.Timestamp

/**
 * Entity to record a particular that a particular achievement has been earned by a player.
 *
 * Points at a GameId if applicable, and uses DtCreation as the "date unlocked".
 */
class AchievementEntity : AbstractEntity<AchievementEntity>()
{
    //DB Fields
    var playerId: Long = -1
    var achievementRef = -1
    var gameIdEarned: Long = -1
    var achievementCounter = -1
    var achievementDetail = ""

    override fun getTableName(): String
    {
        return "Achievement"
    }

    override fun getCreateTableSqlSpecific(): String
    {
        return ("PlayerId INT NOT NULL, "
                + "AchievementRef INT NOT NULL, "
                + "GameIdEarned INT NOT NULL, "
                + "AchievementCounter INT NOT NULL, "
                + "AchievementDetail VARCHAR(255) NOT NULL")
    }

    @Throws(SQLException::class)
    override fun populateFromResultSet(entity: AchievementEntity, rs: ResultSet)
    {
        entity.playerId = rs.getLong("PlayerId")
        entity.achievementRef = rs.getInt("AchievementRef")
        entity.gameIdEarned = rs.getLong("GameIdEarned")
        entity.achievementCounter = rs.getInt("AchievementCounter")
        entity.achievementDetail = rs.getString("AchievementDetail")
    }

    @Throws(SQLException::class)
    override fun writeValuesToStatement(statement: PreparedStatement, startIndex: Int, emptyStatement: String): String
    {
        var i = startIndex
        var statementStr = emptyStatement

        statementStr = writeLong(statement, i++, playerId, statementStr)
        statementStr = writeInt(statement, i++, achievementRef, statementStr)
        statementStr = writeLong(statement, i++, gameIdEarned, statementStr)
        statementStr = writeInt(statement, i++, achievementCounter, statementStr)
        statementStr = writeString(statement, i, achievementDetail, statementStr)

        return statementStr
    }

    override fun addListsOfColumnsForIndexes(indexes: MutableList<MutableList<String>>)
    {
        val ix = HandyArrayList.factoryAdd("PlayerId", "AchievementRef")

        indexes.add(ix)
    }

    companion object
    {

        @JvmStatic fun retrieveAchievement(achievementRef: Int, playerId: Long): AchievementEntity?
        {
            return AchievementEntity().retrieveEntity("PlayerId = $playerId AND AchievementRef = $achievementRef")
        }

        /**
         * Methods for gameplay logic to update achievements
         */
        @JvmStatic fun updateAchievement(achievementRef: Int, playerId: Long, gameId: Long, counter: Int)
        {
            val existingAchievement = retrieveAchievement(achievementRef, playerId)

            if (existingAchievement == null)
            {
                AchievementEntity.factoryAndSave(achievementRef, playerId, gameId, counter)

                triggerAchievementUnlock(-1, counter, achievementRef, playerId, gameId)
            }
            else
            {
                val existingCounter = existingAchievement.achievementCounter

                val decreasing = getAchievementForRef(achievementRef)!!.isDecreasing()

                //Update the achievement if appropriate
                if (counter > existingCounter && !decreasing
                 || counter < existingCounter && decreasing)
                {
                    existingAchievement.achievementCounter = counter
                    existingAchievement.gameIdEarned = gameId
                    existingAchievement.saveToDatabase()

                    triggerAchievementUnlock(existingCounter, counter, achievementRef, playerId, gameId)
                }
            }
        }

        @JvmStatic fun incrementAchievement(achievementRef: Int, playerId: Long, gameId: Long, amountBy: Int = 1)
        {
            val existingAchievement = retrieveAchievement(achievementRef, playerId)

            if (existingAchievement == null)
            {
                AchievementEntity.factoryAndSave(achievementRef, playerId, -1, amountBy)

                triggerAchievementUnlock(-1, amountBy, achievementRef, playerId, gameId)
            }
            else
            {
                val existingCount = existingAchievement.achievementCounter
                existingAchievement.achievementCounter = existingCount + amountBy
                existingAchievement.saveToDatabase()

                triggerAchievementUnlock(existingCount, existingCount + amountBy, achievementRef, playerId, gameId)
            }
        }

        fun insertAchievement(achievementRef: Int, playerId: Long, gameId: Long, detail: String = "")
        {
            val sql = "SELECT COUNT(1) FROM Achievement WHERE PlayerId = $playerId AND AchievementRef = $achievementRef"
            val count = DatabaseUtil.executeQueryAggregate(sql)

            factoryAndSave(achievementRef, playerId, gameId, -1, detail)
            triggerAchievementUnlock(count, count + 1, achievementRef, playerId, gameId)
        }

        private fun triggerAchievementUnlock(oldValue: Int, newValue: Int, achievementRef: Int, playerId: Long, gameId: Long)
        {
            val achievementTemplate = getAchievementForRef(achievementRef)
            triggerAchievementUnlock(oldValue, newValue, achievementTemplate!!, playerId, gameId)
        }

        fun triggerAchievementUnlock(oldValue: Int, newValue: Int, achievementTemplate: AbstractAchievement, playerId: Long, gameId: Long)
        {
            //Work out if the threshold has changed
            achievementTemplate.attainedValue = oldValue

            val oldColor = achievementTemplate.getColor(false)

            achievementTemplate.attainedValue = newValue
            val newColor = achievementTemplate.getColor(false)

            if (oldColor !== newColor)
            {
                //Hooray we've done a thing!
                val scrn = ScreenCache.getDartsGameScreen(gameId)
                scrn.achievementUnlocked(gameId, playerId, achievementTemplate)
            }
        }

        @JvmOverloads
        @JvmStatic fun factoryAndSave(achievementRef: Int, playerId: Long, gameId: Long, counter: Int,
                           achievementDetail: String = "", dtLastUpdate: Timestamp = getSqlDateNow()): AchievementEntity
        {
            val ae = AchievementEntity()
            ae.assignRowId()
            ae.achievementRef = achievementRef
            ae.playerId = playerId
            ae.gameIdEarned = gameId
            ae.achievementCounter = counter
            ae.achievementDetail = achievementDetail
            ae.saveToDatabase(dtLastUpdate)

            return ae
        }
    }
}
